lib/pipelineTemplateVersionFactory.js

```diff
@@ -203,38 +204,48 @@ class PipelineTemplateVersionFactory extends BaseFactory {
     /**
      * Get a version of a pipeline template and template meta based on the criteria
      * @method getWithMetadata
-     * @param  {Object}     config               Config object
-     * @param  {Datastore}  config.datastore     Object that will perform operations on the datastore
-     * @param  {String}     [config.templateId]  The template id
-     * @param  {String}     config.name          The template name
-     * @param  {String}     config.namespace     The template namespace
-     * @param  {String}     config.version       The template version
-     * @param  {Factory}   templateMetaFactory  Template meta factory
-     * @return {Promise}                         {PipelineTemplateVersionModel, PipelineTemplateMetaModel}
+     * @param  {Object}     config                Config object
+     * @param  {String}     config.name           The template name
+     * @param  {String}     config.namespace      The template namespace
+     * @param  {String}     [config.version]      The template version
+     * @param  {String}     [config.versionOrTag] The template version or tag
+     * @param  {Factory}    templateMetaFactory   Template meta factory
+     * @param  {Factory}    [templateTagFactory]  Template tag factory
+     * @return {Promise}                          {PipelineTemplateVersionModel, PipelineTemplateMetaModel}
      */
-    async getWithMetadata(config, templateMetaFactory) {
+    async getWithMetadata(config, templateMetaFactory, templateTagFactory) {
         let pipelineTemplateMeta = {};
+        let templateVersion = config.version;
 
-        if (config.templateId) {
-            pipelineTemplateMeta = await templateMetaFactory.get({
-                id: config.templateId
-            });
-        } else {
-            pipelineTemplateMeta = await templateMetaFactory.get({
+        pipelineTemplateMeta = await templateMetaFactory.get({
+            name: config.name,
+            namespace: config.namespace
+        });
+
+        if (!pipelineTemplateMeta) {
+            return null;
+        }
```

Comment:
Factory should not be designed for a specific (current) need.
We should allow accepting `templateId`.

```suggestion
        const {templateId, name, namespace, versionOrTag, ...versionGetConfig} = config;
        const metaGetConfig = {};
        if(name || namespace) {
            if(name && namespace) {
              metaGetConfig.id = name;
              metaGetConfig.id = namespace;
            }
            else {
               return null; // or throw error
            }
        }
        
        if(templateId) {
            metaGetConfig.id = templateId;
        }
        
        pipelineTemplateMeta = await templateMetaFactory.get(metaGetConfig);

        if (!pipelineTemplateMeta) {
            return null;
        }
        
        versionGetConfig.templateId = pipelineTemplateMeta.id;
```